src/rust/ensogl/lib/core/src/display/symbol/gpu.rs

```diff
@@ -284,31 +284,49 @@ impl Symbol {
         let init_logger = logger.clone();
         debug!(init_logger, "Initializing.", || {
             let on_mut2           = on_mut.clone();
-            let surface_logger    = Logger::sub(&logger,"surface");
-            let shader_logger     = Logger::sub(&logger,"shader");
-            let geo_dirt_logger   = Logger::sub(&logger,"surface_dirty");
-            let mat_dirt_logger   = Logger::sub(&logger,"shader_dirty");
+            let surface_logger    = Logger::new_sub(&logger,"surface");
+            let shader_logger     = Logger::new_sub(&logger,"shader");
+            let geo_dirt_logger   = Logger::new_sub(&logger,"surface_dirty");
+            let mat_dirt_logger   = Logger::new_sub(&logger,"shader_dirty");
             let surface_dirty     = GeometryDirty::new(geo_dirt_logger,Box::new(on_mut2));
             let shader_dirty      = ShaderDirty::new(mat_dirt_logger,Box::new(on_mut));
             let surface_on_mut    = Box::new(f!(surface_dirty.set()));
             let shader_on_mut     = Box::new(f!(shader_dirty.set()));
             let shader            = Shader::new(shader_logger,stats,shader_on_mut);
             let surface           = Mesh::new(surface_logger,stats,surface_on_mut);
-            let variables         = UniformScope::new(Logger::sub(&logger,"uniform_scope"));
+            let variables         = UniformScope::new(Logger::new_sub(&logger,"uniform_scope"));
             let global_variables  = global_variables.clone_ref();
             let bindings          = default();
             let stats             = SymbolStats::new(stats);
             let context           = default();
             let symbol_id_uniform = variables.add_or_panic("symbol_id",(*id) as i32);
             let display_object    = display::object::Instance::new(logger.clone());
             let is_hidden         = Rc::new(Cell::new(false));
-            display_object.set_on_hide(f_!(is_hidden.set(true)));
-            display_object.set_on_show(f__!(is_hidden.set(false)));
             Self {id,display_object,surface,shader,surface_dirty,shader_dirty,variables
-                 ,global_variables,symbol_id_uniform,context,logger,bindings,stats,is_hidden}
+                 ,global_variables,symbol_id_uniform,context,logger,bindings,stats,is_hidden}.init()
         })
     }
 
+    fn init(self) -> Self {
+        let is_hidden = &self.is_hidden;
+        let id        = self.id;
+        self.display_object.set_on_hide(f_!(is_hidden.set(true)));
+        self.display_object.set_on_show(f__!(is_hidden.set(false)));
+        self.display_object.set_on_scene_layer_changed(move |_,old_layers,new_layers| {
+            for layer in old_layers {
+                if let Some(layer) = layer.upgrade() {
+                    layer.remove_symbol(id)
+                }
+            }
```

Comment:
What about
```suggestion
            for layer in old_layers.iter().filter_map(Weak::upgrade) {
                layer.remove_symbol(id)
            }
```
For me it looks a little bit better, but I'm ok with your version too.

Reply:
I like it a lot!